Move PendingTypeLoadTable from the ClassLoader object to a global structure

davidwrighton opened the pull request:
- Shard the hashtable so that it will rarely have lock contention
- Pre-allocate the PendingTypeLoad Entries. This reduces allocator pressure on startup substantially, especially in the presence of multithreaded loading where the struct is allocated on 1 thread and often freed on another.

The effectiveness of the pre-allocation and sharding heuristics was measured on a complex ASP.NET scenario tweaked to perform extremely high numbers of multithreaded loads and produced startup wins of about 10%.

davidwrighton committed:
Move PendingTypeLoadTable from the ClassLoader object to a global structure
- Shard the hashtable so that it will rarely have lock contention
- Pre-allocate the PendingTypeLoad Entries. This reduces allocator pressure on startup substantially, especially in the presence of multithreaded loading where the struct is allocated on 1 thread and often freed on another.

The effectiveness of the pre-allocation and sharding heuristics was measured on a complex ASP.NET scenario tweaked to perform extremely high numbers of multithreaded loads and produced startup wins of about 10%.

AaronRobinsonMSFT commented on src/coreclr/vm/crst.h:
```diff
@@ -105,6 +105,8 @@ extern DWORD g_fEEShutDown;
 extern Volatile<LONG> g_ShutdownCrstUsageCount;
 extern Volatile<LONG> g_fForbidEnterEE;
 
+class PendingTypeLoadTable;
+
```
This shouldn't be necessary. A `friend` declaration in the manner below is sufficient.

davidwrighton replied:
Got rid of that

AaronRobinsonMSFT commented on src/coreclr/vm/pendingload.cpp:
```diff
@@ -12,194 +12,26 @@
-/*static */ PendingTypeLoadTable* PendingTypeLoadTable::Create(LoaderHeap *pHeap,
-                                                               DWORD dwNumBuckets,
-                                                               AllocMemTracker *pamTracker)
+LONG pendingTypeLoadEntryDynamicAllocations = 0;
```
```suggestion
static LONG pendingTypeLoadEntryDynamicAllocations = 0;
```

AaronRobinsonMSFT commented on src/coreclr/vm/pendingload.cpp:
```diff
@@ -12,194 +12,26 @@
 }
+#endif // PENDING_TYPE_LOAD_TABLE_STATS
 
+static BYTE s_PendingTypeLoadTable[sizeof(PendingTypeLoadTable)];
```
This is really an anti-pattern and can create UB easily. Looking at the type there shouldn't be any reason to not declare as:

```cpp
static PendingTypeLoadTable s_PendingTypeLoadTable;
```

davidwrighton replied:
I'm pretty sure that `Crst` types can't be in static variables, and they are all over this thing. However, I will check before my next update.

jkotas replied:
We have `CrstStatic` that is a destructor-free variant of Crst that is meant to be used for static variables.

davidwrighton replied:
Ah, but most of the `Crst` instances are on the Entry structures. Those are structures which both are allocated statically, as well as dynamically in rare cases. I think that makes `CrstStatic` invalid for those scenarios. However, `CrstStatic` could be used for the variables in the PendingTypeLoadTable, and I'll look into if we can do something about the `Entry` structs.

davidwrighton replied:
I was able to get it to work. A bit gross, but its fine.

AaronRobinsonMSFT commented on src/coreclr/vm/pendingload.h:
```diff
@@ -17,248 +17,404 @@
-    friend class ClassLoader;       // workaround really need to beef up the API below
+    static void CallCrstEnter(Crst* pCrst)
+    {
+        pCrst->Enter(INDEBUG(Crst::CRST_NO_LEVEL_CHECK));
```
What is the reason for the `CRST_NO_LEVEL_CHECK`?

elinor-fung replied:
I assume this is from:
https://github.com/dotnet/runtime/blob/971ebdbd0fe335c34d440269af8055c6eea87286/src/coreclr/vm/pendingload.h#L54-L60

But I don't think that reasoning really applies to the refactored/static function now?

AaronRobinsonMSFT commented on src/coreclr/vm/pendingload.h:
```diff
@@ -17,248 +17,404 @@
-            THROWS;
-            GC_TRIGGERS;
-            INJECT_FAULT(COMPlusThrowOM(););
+            m_typeKey = typeKey;
```
Is this copy desired? If so, it is more efficient to pass in the argument by value.

AaronRobinsonMSFT commented on src/coreclr/vm/pendingload.h:
```diff
@@ -17,248 +17,404 @@
-        EX_CATCH
+    #endif //DACCESS_COMPILE
+
+        TypeKey& GetTypeKey()
```
Is mutability desired here?

davidwrighton replied:
No, we just want the type key. I've fixed this by making it a const pointer, and then updating the various parts of our codebase to allow a const TypeKey pointer to be used. Looks a little better to me.

AaronRobinsonMSFT commented on src/coreclr/vm/pendingload.h:
```diff
@@ -17,248 +17,404 @@
-    void SetResult(TypeHandle typeHnd)
-    {
-        CONTRACTL
+        // This is only safe to call on an AddRef'd PendingTypeLoadEntry which has had DelayForProgress() called on it
```
Seems like an opportunity for an assert.

davidwrighton replied:
Yes, but asserts with distributed thread local state are a real headache to write. Instead, I'll adjust `DelayForProgress` to return the `m_typeHandle` when appropriate.

AaronRobinsonMSFT commented on src/coreclr/vm/pendingload.h:
```diff
@@ -17,248 +17,404 @@
+                    // We won't be holding the lock while Releasing, so use a VolatileStore to ensure all writes during Reset are complete.
+                    VolatileStore(&m_fIsUnused, true);
+                }
+                else
+                    delete this;
```
```suggestion
                else
                {
                    delete this;
                }
```

AaronRobinsonMSFT commented on src/coreclr/vm/pendingload.h:
```diff
@@ -17,248 +17,404 @@
-        m_fLockAcquired = FALSE;
-    }
-#endif //DACCESS_COMPILE
+        BOOL HasWaiters()
```
```suggestion
        bool HasWaiters()
```

AaronRobinsonMSFT commented on src/coreclr/vm/pendingload.h:
```diff
@@ -17,248 +17,404 @@
+
+            Entry **pCurrent = &m_pLinkedListOfActiveEntries;
+
+            while (*pCurrent != pEntry)
+            {
+                pCurrent = &((*pCurrent)->m_pNext);
+            }
+            *pCurrent = (*pCurrent)->m_pNext;
```
I don't see the case being handled where `pEntry` isn't in the list.

davidwrighton replied:
That isn't a valid case for calling this method. The `PendingTypeLoadEntry` *must* be in the list before calling this api. I'll add a comment.

AaronRobinsonMSFT commented on src/coreclr/vm/pendingload.h:
```diff
@@ -17,248 +17,404 @@
 
 #ifdef _DEBUG
-    DWORD           m_dwDebugMemory;
+        void Dump();
```
It is nice that `Dump()` in the compilation unit. It would be nice if more of the above implementation was as well.

AaronRobinsonMSFT commented on src/coreclr/vm/pendingload.h:
```diff
@@ -17,248 +17,404 @@
 
-    static PendingTypeLoadTable *Create(LoaderHeap *pHeap, DWORD dwNumBuckets, AllocMemTracker *pamTracker);
+    // This number chosen by experimentation with a fairly complex ASP.NET application that would naturally use about 40,000 Entry structures on startup.
+    // Entry allocations were shifted to about 11 during that startup phase.
```
How does 11 relate to 31?

davidwrighton replied:
Ah, its an unclear comment... we previously had needed to allocate and free 40,000 or so different Entry structs, and with the new code in that environment we needed to allocate around 11. (The exact number fluctuated based on which threads made progress in what order, so it wasn't a completely stable number, but that's expected.)

davidwrighton committed:
First round of fixes

davidwrighton committed:
Address more feedback
- Use a const TypeKey
  - Required making a fair amount of other code const correct
- Implement in cpp file instead of header
- Don't use initialization trick

AaronRobinsonMSFT commented on src/coreclr/src/coreclr:
Is this needed?

davidwrighton replied:
No... what the heck made this thing. Sigh.

AaronRobinsonMSFT commented on src/coreclr/vm/pendingload.h:
```diff
@@ -20,245 +20,150 @@
-        PendingTypeLoadEntry* pData;
-    };
+private:
+        Shard()
+        {}
```
```suggestion
        Shard() = default;
```

davidwrighton committed:
Code review updates

AaronRobinsonMSFT commented on src/coreclr/vm/pendingload.cpp:
```diff
@@ -12,241 +12,365 @@
+#endif // PENDING_TYPE_LOAD_TABLE_STATS
+}
+
+void PendingTypeLoadTable::Entry::SetTypeKey(const TypeKey& typeKey)
```
```suggestion
void PendingTypeLoadTable::Entry::SetTypeKey(TypeKey typeKey)
```

This should be okay.

AaronRobinsonMSFT commented on src/coreclr/vm/pendingload.cpp:
```diff
@@ -12,241 +12,365 @@
     }
 
-    return FALSE;
+    return m_hrResult;
```
```suggestion
    return hr;
```

I assume the logic here is we want to return the operation we branched on. I'm being defensive for non-obvious races.

davidwrighton committed:
Hopefully final set of updates
